web_src/js/features/repo-common.js

```diff
@@ -1,44 +1,42 @@
 import $ from 'jquery';
-import {hideElem, showElem} from '../utils/dom.js';
+import {hideElem, queryElems, showElem} from '../utils/dom.js';
 import {POST} from '../modules/fetch.js';
+import {showErrorToast} from '../modules/toast.js';
 
-async function getArchive($target, url, first) {
-  const dropdownBtn = $target[0].closest('.ui.dropdown.button') ?? $target[0].closest('.ui.dropdown.btn');
-
+async function downloadArchive(target, url, retryCount = 0) {
+  // there are many places using the "archive-link", eg: the dropdown on the repo code page, the release list
+  const targetLoading = target.closest('.ui.dropdown') ?? target;
+  let keepLoading = false;
   try {
-    dropdownBtn.classList.add('is-loading');
+    targetLoading.classList.add('is-loading', 'loading-icon-2px');
     const response = await POST(url);
-    if (response.status === 200) {
-      const data = await response.json();
-      if (!data) {
-        // XXX Shouldn't happen?
-        dropdownBtn.classList.remove('is-loading');
-        return;
-      }
-
-      if (!data.complete) {
-        // Wait for only three quarters of a second initially, in case it's
-        // quickly archived.
-        setTimeout(() => {
-          getArchive($target, url, false);
-        }, first ? 750 : 2000);
-      } else {
-        // We don't need to continue checking.
-        dropdownBtn.classList.remove('is-loading');
-        window.location.href = url;
-      }
+    if (!response.ok) {
+      throw new Error(`Invalid server response: ${response.status}`);
+    }
+    const data = await response.json();
+    if (!data.complete) {
+      keepLoading = true; // the archive is not ready yet, keep loading and then retry later
+      const delay = Math.min((retryCount + 1) * 750, 2000);
+      setTimeout(() => downloadArchive(target, url, retryCount + 1), delay);
+    } else {
+      window.location.href = url; // the archive is ready, start real downloading
     }
-  } catch {
-    dropdownBtn.classList.remove('is-loading');
+  } catch (e) {
+    console.error(e);
+    showErrorToast(`Failed to download the archive: ${e}`, {duration: 2500});
+  } finally {
+    if (!keepLoading) targetLoading.classList.remove('is-loading', 'loading-icon-2px');
   }
 }
 
 export function initRepoArchiveLinks() {
-  $('.archive-link').on('click', function (event) {
-    event.preventDefault();
-    const url = this.getAttribute('href');
-    if (!url) return;
-    getArchive($(event.target), url, true);
+  queryElems('.archive-link', (el) => {
+    el.addEventListener('click', (e) => {
+      const target = e.target.closest('.archive-link');
+      if (!target?.href) return;
+      e.preventDefault();
+      downloadArchive(target, target.href);
+    });
```

Comment:
Hmm, if we already have a big function body, I think I prefer `for (const el of document.querySelectorAll)`, because of the benefits of being able to `break` and `continue`.

Reply:
1. I do not think it really needs "break"
2. It could just "return" to ignore current element like "continue"
3. I think it is more readable than a for-loop. If you prefer, feel free to rewrite it to see the difference.

Reply:
There are more cases for `queryElems` like this:

```
queryElems('.selector', (el) => el.disabled = true);
```

It would make the jQuery rewriting easier, and less bug.

Reply:
For single-line stuff, I agree, such function is nice.

Generally thought, I prefer to avoid callback in favor of for-of, reasons being the unavailable flow control statements and there are more problems like [one can not yield from a callback](https://stackoverflow.com/questions/41326217/js-how-to-use-generator-and-yield-in-a-callback).

Reply:
I do not see any real problem, it just works like `queryElemsChildren` and `queryElemsSiblings`

I would strongly prefer to introduce it, to help migrating from jQuery

Reply:
> Generally thought, I prefer to avoid callback in favor of for loops, reasons being the unavailable flow control statements and there are more problems like [one can not yield from a callback](https://stackoverflow.com/questions/41326217/js-how-to-use-generator-and-yield-in-a-callback).

The question is: why you need to do so? What's the real use case with `queryElems`? Even if it is a case, it could be simply rewritten to `for (const el of queryElems('.selector'))`, it is still shorter to read.